ebms-async/src/main/kotlin/no/nav/emottak/ebms/async/kafka/consumer/FailedMessageKafkaHandler.kt

```diff
@@ -0,0 +1,173 @@
+
+    suspend fun send(record: ReceiverRecord<String, ByteArray>, key: String = record.key(), value: ByteArray = record.value()) {
+        record.addHeader(RETRY_AFTER, getNextRetryTime(record))
+        try {
+            val result = producer.send(ProducerRecord(kafkaErrorQueue.topic, null, key, value, record.headers())).get()
+            logger.info("Wrote to offset:" + result.offset())
+            producer.close()
+//            producersFlow.collect { producer ->
+//                val metadata = producer.send(ProducerRecord(kafkaErrorQueue.topic, null, key, value, record.headers())).get()
+//                logger.info("Offset on metadata: " + metadata.offset())
+//                logger.info("Result " + metadata.partition() + " timestamp " + metadata.timestamp())
+//                producer.commitTransaction()
+//                producer.close()
+//            }
+            logger.info("Message sent successfully to topic ${kafkaErrorQueue.topic}")
+        } catch (e: Exception) {
+            logger.info("Failed to send message to ${kafkaErrorQueue.topic} : ${e.message}")
+        }
+    }
+
+    suspend fun receive(payloadMessageProcessor: PayloadMessageProcessor, limit: Int = 10) { // TODO limit til offset
+        logger.debug("Reading from error queue")
+        var counter = 0
+        consumerFlow.map { record ->
+            counter++
+            if (counter > limit) {
+                throw Exception("Error queue limit exceeded: $limit") // TODO fjern dette
+            }
+            record.offset.acknowledge()
+            record.retryCounter()
+            payloadMessageProcessor.process(record)
+            record.offset.acknowledge()
+        }.collect()
+    }
+
+    fun getNextRetryTime(record: ReceiverRecord<String, ByteArray>): String {
+        return DateTime.now().plusMinutes(5)
+            .toString() // TODO create retry strategy
+    }
+
+    fun ReceiverRecord<String, ByteArray>.retryCounter(): BigInteger {
+        val lastHeader = headers().lastHeader(RETRY_COUNT_HEADER)?.value() ?: (0).toBigInteger().toByteArray()
+        val retryCounter = BigInteger(lastHeader) + (1).toBigInteger()
+        this.headers().add(
+            RETRY_COUNT_HEADER,
+            retryCounter.toByteArray()
+        )
+        return retryCounter
+    }
+}
+
+fun ReceiverRecord<String, ByteArray>.addHeader(key: String, value: String) {
+    this.headers().add(key, value.toByteArray())
+}
+
+fun getRetryRecord(fromOffset: Long = 0, requestedRecords: Int = 1): ReceiverRecord<String, ByteArray>? {
+    return getRecord(config().kafkaErrorQueue.topic, config().kafka, fromOffset, requestedRecords)
+}
+
+fun getRecord(topic: String, kafka: Kafka, fromOffset: Long = 0, requestedRecords: Int = 1): ReceiverRecord<String, ByteArray>? {
+    return with(
+        KafkaConsumer(
+            kafka.toProperties(),
+            StringDeserializer(),
+            ByteArrayDeserializer()
+        )
+    ) {
+        partitionsFor(topic)
```

Comment:
Det er mye som foregår her, hva med å splitte dette opp i et par mindre funksjoner for å tydeliggjøre hva som faktisk skjer